### PR TITLE
feat(aws-s3): adds s3 bucket aws fsbp option

### LINKED ISSUE #10969: [aws-s3] Enforce AWS Foundational Security Best Practice
It would be nice to have the ability to enforce [AWS Foundational Security Best Practice](https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html) through CDK.

### Use Case

When creating s3 buckets AWS FSBP should be followed.

### Proposed Solution

```
      new s3.Bucket(stack, 'MyBucket', {
        enforceSecurityBestPractice: true,
      });
```
And behind the scenes it will block all public access and enforce data transmission over HTTPS.

A PR is ready to go which introduces a simple way of adding these controls for s3 buckets.

### Other

This is really just to get the conversation started around enforcing the AWS FSBP controls as a default for CDK and perhaps this could be a first step into helping prevent AWS users from using defaults which are not secure. 

* [X] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by crashGoBoom**
This adds an option to enforce aws foundational best practices for s3 buckets.

Closes #10969
Signed-off-by: Christopher Mundus <chris@kindlyops.com>

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by NetaNir** [comment]
Hi @ crashGoBoom! 

Generally we prefer to encode best practices as defaults in the L2 constructs. Individual properties, such as the the `enforceSsl`, should be exposed as properties and have sane defaults, allowing users to opt out from specific configurations. If you agree, I suggest adding `enforceSsl` as property and removing the `enforceSecurityBestPractice` property.

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 1]
```diff
@@ -1392,6 +1413,17 @@ export class Bucket extends BucketBase {
     this.inventories.push(inventory);
   }
 
+  private enforceSSL() {
+    const statement = new iam.PolicyStatement({
+      actions: ['s3:*'],
+      effect: iam.Effect.DENY,
+      resources: [this.bucketArn, `${this.bucketArn}/*`],
```
Comment (NetaNir) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts:1420]:
```suggestion
      resources: [this.arnForObjects('*')]
```

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 2]
```diff
@@ -1392,6 +1413,17 @@ export class Bucket extends BucketBase {
...
+      principals: [new iam.AnyPrincipal()],
+    });
+    statement.addCondition('Bool', { 'aws:SecureTransport': 'false' });
```
Comment (NetaNir) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts:1423]: Can be configured on the constructor:

```
conditions: {
  Bool: { 'aws:SecureTransport': 'false' },
  },

```

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 3]
```diff
@@ -1012,6 +1012,14 @@ export interface BucketProps {
    */
   readonly encryptionKey?: kms.IKey;
 
+  /**
+  * Enforces all of the AWS Foundational Security Best Practices Regarding S3
+  * Details: https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html
+  *
+  * @default false
+  */
+  readonly enforceSecurityBestPractice?: boolean;
```
Comment (NetaNir) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts:1021]: Lets remove this, see my comment on the PR.

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 4]
```diff
@@ -1392,6 +1413,17 @@ export class Bucket extends BucketBase {
     this.inventories.push(inventory);
   }
 
+  private enforceSSL() {
```
Comment (NetaNir) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts:1416]: Expose `enforceSSL` as a property

**Review by NetaNir — DISMISSED** [review]

File: packages/@aws-cdk/aws-s3/lib/bucket.ts (hunk 4 above)
Comment (crashGoBoom) [reply packages/@aws-cdk/aws-s3/lib/bucket.ts]: Ok I just removed the enforceSecurityBestPractice property and replaced it with enforceSSL. Thanks for the input! Let me know if there is anything else.

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 5]
```diff
@@ -1012,6 +1012,14 @@ export interface BucketProps {
    */
   readonly encryptionKey?: kms.IKey;
 
+  /**
+  * Enforces SSL for requests. S3.5 of the AWS Foundational Security Best Practices Regarding S3.
+  * @see https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html#fsbp-s3-5
```
Comment (NetaNir) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts:1017]: replace with the docs specific to s3: https://docs.aws.amazon.com/config/latest/developerguide/s3-bucket-ssl-requests-only.html

**Review by NetaNir — CHANGES_REQUESTED** [review]
Looks good, one comment about the doc. Feel free to tag me so I can merge this, thanks!

File: packages/@aws-cdk/aws-s3/lib/bucket.ts (hunk 5 above)
Comment (crashGoBoom) [reply packages/@aws-cdk/aws-s3/lib/bucket.ts]: @NetaNir Hi! I just updated it. Thanks so much!

**Comment by crashGoBoom** [comment]
> Looks good, one comment about the doc. Feel free to tag me so I can merge this once updated! Thanks!

@NetaNir Hi! This has been updated. Let me know if there is anything else. Thanks so much!

**Comment by NetaNir** [comment]
@mergifyio update

**Comment by NetaNir** [comment]
Hi @crashGoBoom see mergify comment above, you will need to update your branch so I can merge it.

**Comment by crashGoBoom** [comment]
> Hi @crashGoBoom see mergify comment above, you will need to update your branch so I can merge it.

Ok I just updated sorry about that! @NetaNir

File: packages/@aws-cdk/aws-s3/test/bucket.test.ts [hunk 6]
```diff
@@ -495,6 +495,62 @@ nodeunitShim({
       test.done();
     },
 
+    'bucket with aws foundational security best practice enforce ssl'(test: Test) {
```
Comment (NetaNir) [inline packages/@aws-cdk/aws-s3/test/bucket.test.ts:498]:
```suggestion
    'can set enforceSsl'(test: Test) {
```

**Review by NetaNir — DISMISSED** [review]

File: packages/@aws-cdk/aws-s3/test/bucket.test.ts (hunk 6 above)
Comment (NetaNir) [reply packages/@aws-cdk/aws-s3/test/bucket.test.ts:498]: Sorry for the many rounds :/ you can commit this one from theGitHub UI

Comment (crashGoBoom) [reply packages/@aws-cdk/aws-s3/test/bucket.test.ts]: @NetaNir no worries! Ok I just commited that change.

**Comment by crashGoBoom** [comment]
@NetaNir ok looks like it should be ready to go now.

**Review by NetaNir — APPROVED** [review]

**Comment by crashGoBoom** [comment]
Hmm, looks like this might have failed to allow access since this PR is coming from an ORG repo instead of a user repo. I am not seeing the options to allow changes to be pushed to this fork so I am guessing thats what it is. @NetaNir Any ideas if its possible to use this org fork or should this PR be closed and opened with my user instead of our org? Thanks and sorry for the hassle!

**Comment by iliapolo** [comment]
Closed in favor of https://github.com/aws/aws-cdk/pull/12804

**Pull request closed without merging**